### PR TITLE
warnings: fix code that emits compiler warnings

### PR DESCRIPTION
**Pull request by vtjnash**
Added `-Wno-long-long` on Windows to disable warning about ULL constants (in the filetime code). This flag was already being used for the test suite.

Not many real bugs were uncovered, but does make compilation a bit less noisy.

### CODE REVIEW

**Comment by vtjnash** [comment]
Actually, I might need help with the AIX warnings from someone with access to the platform:

```
  CC       src/unix/libuv_la-core.lo
src/unix/core.c: In function 'uv__nonblock_ioctl':
src/unix/core.c:551:5: warning: overflow in implicit constant conversion [-Woverflow]
     r = ioctl(fd, FIONBIO, &set);
     ^
```
```
  CC       src/unix/libuv_la-fs.lo
src/unix/fs.c: In function 'uv__fs_mkdtemp':
src/unix/fs.c:219:3: warning: implicit declaration of function 'mkdtemp' [-Wimplicit-function-declaration]
   return mkdtemp((char*) req->path) ? 0 : -1;
   ^
```

**Comment by richardlau** [comment]
> ```
>   CC       src/unix/libuv_la-fs.lo
> src/unix/fs.c: In function 'uv__fs_mkdtemp':
> src/unix/fs.c:219:3: warning: implicit declaration of function 'mkdtemp' [-Wimplicit-function-declaration]
>    return mkdtemp((char*) req->path) ? 0 : -1;
>    ^
> ```

https://github.com/libuv/libuv/pull/740

**Comment by richardlau** [comment]
> 
> 
> Actually, I might need help with the AIX warnings from someone with access to the platform:
> 
> ```
>   CC       src/unix/libuv_la-fs.lo
> src/unix/fs.c: In function 'uv__fs_mkdtemp':
> src/unix/fs.c:219:3: warning: implicit declaration of function 'mkdtemp' [-Wimplicit-function-declaration]
>    return mkdtemp((char*) req->path) ? 0 : -1;
>    ^
> ```

https://github.com/libuv/libuv/pull/740. From the linked issue (https://sourceware.org/bugzilla/show_bug.cgi?id=13558#c1):

> Exactly: Since AIX 6.1, mkdtemp() is declared upon _XOPEN_SOURCE >= 700, and thus the symbol is available in libc.
> 
> The difference is: Upon _ALL_SOURCE (defined in config.h),
> AIX 6.1 defines (overrides!) _XOPEN_SOURCE=600, while
> AIX 7.1 defines (overrides!) _XOPEN_SOURCE=700, 
> 
> This particularly isn't a problem on AIX 7.1, but still the check for mkdtemp() should be extended to if the declaration is available too.

Unfortunately the CI machines are AIX 6.1.

cc @gdams @gireeshpunathil @mhdawson

File: test/test-condvar.c [hunk 1]
```diff
@@ -235,7 +235,7 @@ TEST_IMPL(condvar_5) {
   uint64_t elapsed;
   uint64_t timeout;
 
-  timeout = 100 * 1e6; /* 100 ms in ns */
+  timeout = 100e6; /* 100 ms in ns */
```
Comment (refack) [inline test/test-condvar.c]: `100e6` is still a double literal.
```suggestion
  timeout = 100 * 1000 * 1000; /* 100ms in ns */
```

File: uv.gyp [hunk 2]
```diff
@@ -97,6 +97,7 @@
             '_WIN32_WINNT=0x0600',
             '_GNU_SOURCE',
           ],
+          'cflags': [ '-Wno-long-long' ],
```
Comment (refack) [inline uv.gyp]: Has no effect on MSVS, and not needed after tweak to `test/test-condvar.c`

File: Makefile.am [hunk 3]
```diff
@@ -51,6 +51,7 @@ uvinclude_HEADERS += include/uv/win.h include/uv/tree.h
 AM_CPPFLAGS += -I$(top_srcdir)/src/win \
                -DWIN32_LEAN_AND_MEAN \
                -D_WIN32_WINNT=0x0600
+libuv_la_CFLAGS += -Wno-long-long
```
Comment (refack) [inline Makefile.am]: Has no effect on MSVS, and not needed after tweak to `test/test-condvar.c`

**Comment by refack** [comment]
MSVS building for 32bit is clean with this PR + the tweak in `test/test-condvar.c`.
64bit is still has 65 warnings (some might be legit, like `'ULONG_PTR' to 'DWORD'`)

File: src/win/dl.c [hunk 4]
```diff
@@ -64,7 +64,7 @@ void uv_dlclose(uv_lib_t* lib) {
 
 
 int uv_dlsym(uv_lib_t* lib, const char* name, void** ptr) {
-  *ptr = (void*) GetProcAddress(lib->handle, name);
+  *ptr = (void*)(uintptr_t) GetProcAddress(lib->handle, name); /* Cast though integer to suppress pedantic warning about forbidden cast. */
```
Comment (refack) [inline src/win/dl.c]: I don't see MSVS complaining (I also don't think it should). so maybe leave it alone?

File: src/win/winapi.h [hunk 5]
```diff
@@ -4109,6 +4109,7 @@
 #endif
 
 /* from winternl.h */
+#define __UNICODE_STRING_DEFINED
```
[the remark this answers is not in the document]
Comment (refack) [reply src/win/winapi.h]:
```suggestion
#ifndef __UNICODE_STRING_DEFINED
#if !defined (__UNICODE_STRING_DEFINED) && defined (__MINGW32_)
#define __UNICODE_STRING_DEFINED
```

File: src/win/dl.c [hunk 6]
```diff
@@ -75,8 +75,9 @@ const char* uv_dlerror(const uv_lib_t* lib) {
 
 
 static void uv__format_fallback_error(uv_lib_t* lib, int errorno){
-  DWORD_PTR args[1] = { (DWORD_PTR) errorno };
-  LPSTR fallback_error = "error: %1!d!";
+  static const CHAR fallback_error[] = "error: %1!d!";
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/win/dl.c]: The code is required to adhere to C89 syntax according to https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md#code

Yes, C permits this declaration for legacy reasons (ISO C++11 would have rejected it for dropping `const`), but the rewritten type is more accurately the type of the RHS.

File: src/win/dl.c (hunk 4 above)
Comment (vtjnash) [reply src/win/dl.c]: GCC complains, since officially it's UB (although only under -Wpedantic). It's pretty silly, but that's what it is, and this is the only way to disable that warning that I'm aware of.

ref https://stackoverflow.com/questions/36645660/why-cant-i-cast-a-function-pointer-to-void

**Comment by vtjnash** [comment]
> 64bit is still has 65 warnings (some might be legit, like 'ULONG_PTR' to 'DWORD')

I see no warnings on CI. Where do you see these?

File: src/win/winapi.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (vtjnash) [reply src/win/winapi.h]: Sure, but why is that better?

**Comment by vtjnash** [comment]
bump @refack (mentioning additional warnings to fix), and folks who worked on #740 (if my solution proposal is acceptable)

File: src/win/dl.c (hunk 4 above)
Comment (refack) [reply src/win/dl.c]: Ack

File: src/win/dl.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (refack) [reply src/win/dl.c]: Hmmm 🤔 C89... Ok.

File: src/win/fs.c [hunk 7]
```diff
@@ -115,8 +115,8 @@
 
 #define TIME_T_TO_FILETIME(time, filetime_ptr)                              \
   do {                                                                      \
-    uint64_t bigtime = ((uint64_t) ((time) * 10000000ULL)) +                \
-                                  116444736000000000ULL;                    \
+    uint64_t bigtime = ((uint64_t) ((time) * (uint64_t) 10000 * 1000U)) +       \
```
Comment (refack) [inline src/win/fs.c]: I think the continuation `/`s should align for all line of the MACRO

File: src/win/fs.c [hunk 8]
```diff
@@ -115,8 +115,8 @@
...
+                                  (uint64_t) 116444736 * 1000U * 1000U * 1000U; \
```
Comment (refack) [inline src/win/fs.c]: And this line is 81 chars long.

File: src/win/winapi.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (refack) [reply src/win/winapi.h]: Since it "might" have a non empty value already defined. And it seems like the more common pattern used elsewhere...

**Comment by refack** [comment]
AFAIK the mention of C89 in https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md#code is spesific for variable decleration order.
All of our supported compilers know C99, so C89 wasn't enforced or checked. I'm pro modernization, but I defer to the more senior members.

**Comment by mhdawson** [comment]
@refack is AIX still an issue?

**Comment by refack** [comment]
@mhdawson IIUC yes (https://github.com/libuv/libuv/pull/2066#issuecomment-436126491). @vtjnash is asking for feedback if the suggested solution is acceptable (since we can't CI it).

**Comment by bnoordhuis** [comment]
> All of our supported compilers know C99, so C89 wasn't enforced or checked.

Except MSVC. :-) More recent versions do better but we nominally still support MSVC 2008, which does not - it only knows a subset of C99.

That doesn't need to hold back code in src/unix but back then we decided to stick to C89 for consistency. I don't have a problem with switching to a newer dialect, personally.

**Comment by vtjnash** [comment]
Previous discussion (merged, making master now declare c99) is at https://github.com/libuv/libuv/pull/994

Note that while MSVC has been supporting a larger subset of C99 as time goes on, it still doesn't implement all of it (https://docs.microsoft.com/en-us/cpp/visual-cpp-language-conformance?view=vs-2017). In particular, VLA support might never be implemented (they've apparently been removed from the C11 requirements), and the variadic macro implementation is still being overhauled. 

On the other hand, mixed code and declarations was added in VS2013, so that policy could now be altered. Although, as Ben mentioned above, when this policy was adopted, it was done so for consistency in the codebase and not strictly just technical reasons. (saghul mentioned a similar preference in https://github.com/libuv/libuv/pull/994#issuecomment-239382014)

> so C89 wasn't enforced or checked

It is checked (insofar as gcc and msvc implements warnings for it). It's not enforced by Jenkins (it might be good to add `-Werror` to the CI configuration to prevent future regressions), but that's what this PR is addressing.

**Comment by refack** [comment]
> > so C89 wasn't enforced or checked
> 
> It is checked (insofar as gcc and msvc implements warnings for it). It's not enforced by Jenkins (it might be good to add `-Werror` to the CI configuration to prevent future regressions), but that's what this PR is addressing.

Yes, I meant not checked in CI.
`-Werror` and `/W4` or `/WX` could be added to the build scaffolding.

**Comment by vtjnash** [comment]
New CI run: https://ci.nodejs.org/job/libuv-test-commit/1109/

Remaining warnings:

AIX were assigning a function that's supposed to return an `int` to a variable of type `int`, but perhaps the prototype for this function was wrong on AIX 6.1?
```
src/unix/core.c: In function 'uv__nonblock_ioctl':
src/unix/core.c:551:5: warning: overflow in implicit constant conversion [-Woverflow]
     r = ioctl(fd, FIONBIO, &set);
     ^
```

smartos 14 has a couple nuisance warnings, but these go away with newer versions of the OS / compiler:
```
src/unix/sunos.c: In function 'uv_fs_event_start':
src/unix/sunos.c:490:16: warning: 'portfd' may be used uninitialized in this function [-Wmaybe-uninitialized]
     uv__io_init(&handle->loop->fs_event_watcher, uv__fs_event_read, portfd);
                ^
test/test-fork.c: In function 'assert_wait_child':
test/test-fork.c:81:3: warning: format '%d' expects argument of type 'int', but argument 2 has type 'pid_t' [-Wformat=]
   printf("Waited pid is %d with status %d\n", waited_pid, child_stat);
   ^
```

On zos, two casts were missing (now fixed in followup commit):
```
WARNING CCN3280 ../src/unix/fs.c:729   Function argument assignment between types "char*" and "const char*" is not allowed.
WARNING CCN3068 ../src/unix/os390.c:232   Operation between types "struct Output_path_type*" and "char*" is not allowed.
```

and there's few more nuisance warnings in the tests on zos:
```
 WARNING CCN1109: Infinite loop detected in function maybe_run_test.  Program may not stop.
 WARNING CCN1142: NOSTRICT may alter the semantics of a program.
```

File: src/unix/fs.c [hunk 9]
```diff
@@ -70,6 +70,10 @@
 # include <utime.h>
 #endif
 
+#if defined(_AIX) && _XOPEN_SOURCE == 600
```
Comment (bnoordhuis) [inline src/unix/fs.c]: `<=`?

File: src/win/dl.c (hunk 4 above)
Comment (bnoordhuis) [inline src/win/dl.c]: Long line.

File: src/win/fs.c [hunk 10]
```diff
@@ -115,8 +115,8 @@
 
 #define TIME_T_TO_FILETIME(time, filetime_ptr)                              \
   do {                                                                      \
-    uint64_t bigtime = ((uint64_t) ((time) * 10000000ULL)) +                \
-                                  116444736000000000ULL;                    \
+    uint64_t bigtime = ((uint64_t) ((time) * (uint64_t) 10000 * 1000U)) +   \
+                       (uint64_t) 116444736 * 1000U * 1000U * 1000U;        \
```
Comment (bnoordhuis) [inline src/win/fs.c]: Maybe define `MILLION` and `BILLION` macros and use those. Just a suggestion though, this isn't worse than what it replaces.

File: src/win/fs.c [hunk 11]
```diff
@@ -788,8 +788,7 @@ void fs__unlink(uv_fs_t* req) {
     /* Remove read-only attribute */
     FILE_BASIC_INFORMATION basic = { 0 };
 
-    basic.FileAttributes = info.dwFileAttributes
-                           & ~(FILE_ATTRIBUTE_READONLY)
+    basic.FileAttributes = (info.dwFileAttributes & ~FILE_ATTRIBUTE_READONLY)
                            | FILE_ATTRIBUTE_ARCHIVE;
```
Comment (bnoordhuis) [inline src/win/fs.c]: Since you're here, can you put the `|` on the previous line?

File: src/win/fs.c [hunk 12]
```diff
@@ -1201,7 +1200,7 @@ INLINE static int fs__stat_handle(HANDLE handle, uv_stat_t* statbuf,
 
   /* st_blocks contains the on-disk allocation size in 512-byte units. */
   statbuf->st_blocks =
-      file_info.StandardInformation.AllocationSize.QuadPart >> 9ULL;
+      file_info.StandardInformation.AllocationSize.QuadPart >> (uint64_t) 9;
```
Comment (bnoordhuis) [inline src/win/fs.c]: No cast necessary, I think? Plain 9 should suffice.

File: src/win/pipe.c [hunk 13]
```diff
@@ -2141,7 +2141,7 @@ int uv_pipe_open(uv_pipe_t* pipe, uv_file file) {
   if (pipe->ipc) {
     assert(!(pipe->flags & UV_HANDLE_NON_OVERLAPPED_PIPE));
     pipe->pipe.conn.ipc_remote_pid = uv_os_getppid();
-    assert(pipe->pipe.conn.ipc_remote_pid != -1);
+    assert(pipe->pipe.conn.ipc_remote_pid != (DWORD)(uv_pid_t) -1);
```
Comment (bnoordhuis) [inline src/win/pipe.c]: Only needs a cast to DWORD, right?

File: test/test-fork.c [hunk 14]
```diff
@@ -676,5 +676,8 @@ TEST_IMPL(fork_threadpool_queue_work_simple) {
 }
 #endif /* !__MVS__ */
 
+#else
+
+extern int this_file_has_no_tests; /* ISO C forbids an empty translation unit. */
```
Comment (bnoordhuis) [inline test/test-fork.c]: Just over the 80 column limit, I'm afraid.

File: src/win/fs.c (hunk 12 above)
Comment (vtjnash) [reply src/win/fs.c]: I think this is one of the horrible edge cases of the promotion rules where an explicit cast is required (on at least one side of the binary operator) to preserve the exact meaning of the current expression.

Comment (bnoordhuis) [reply src/win/fs.c]: Can you walk me through that? QuadPart is 64 bits, isn't it? If it isn't, I'd expect the cast on the LHS, not the RHS.

Comment (bnoordhuis) [reply src/win/fs.c]: Ah, okay. I didn't realize QuadPart is signed. Can I suggest putting the cast on the LHS for legibility reasons? You can tell it wasn't completely obvious to me so it probably isn't obvious to others either.

**Comment by vtjnash** [comment]
New CI run passing: https://ci.nodejs.org/job/libuv-test-commit/1110/

**Comment by vtjnash** [comment]
> Yes, I meant not checked in CI.

If I wanted to explore enabling this (and adding Win64 CI), where would I find this configuration? Or do I not have access to see it.

**Comment by refack** [comment]
@vtjnash, 👋 I'm your friendly BuildWG member in residence. Either ping me directly or open an issue at https://github.com/nodejs/build.

I'll try to get to this (adding Win64) later today.

Just for reference, we (BuildWG) use Jenkins machine labels to select for 86/64 tests. So for example [`win-vs2017-x86`](https://ci.nodejs.org/label/win-vs2017-x86/) is used as a variable in the script to select for `x86` flavored testing, while [`win2012r2-vs2017`](https://ci.nodejs.org/label/win2012r2-vs2017/) is usually used for default testing, but I think in this case we should use it for `x64` tests.
